### DISCUSSION TITLE
On the front end, I want to use IModelJson.Reader.parse Construct a geometryquery based on geometricelement props data. Can you help me? thank you

### DISCUSSION THREAD

**Discussion by shenshengyi** [iTwin/itwinjs-core#339] [Q&A] [answered]
Hi, I select a geometry on the web, and then get the geometricelementprops data (from backend) according to its ID, and try to pass the IModelJson.Reader.parse To construct a geometryquery object for frontend to use,

But it failed. I see it in the imodeljs test code IModelJson.Reader.parse The data used is either from external files or from JSON format data written by ourselves. Is my idea wrong? Can you help me? thank you

**Comment by pmconne** [comment, marked as answer]
Please be specific (or better, show code):
- How did you obtain GeometricElementProps - `IModelConnection.elements.getProps()`?
- What did you try to pass to `IModelJsonReader.parse()` - `GeometricElementProps.geom`?
- What do you mean by "it failed"? Did it produce exceptions - if so, supply them.

Did you check if your `GeometricElementProps` object's `geom` member was `undefined`?
`IModelConnection.elements.getProps()` omits the geometry because (1) it can be an extremely large amount of data that (2) most callers don't care about and (3) only a subset of it can be processed on the front-end.

What specifically are you trying to accomplish?

**Reply by shenshengyi** [reply]
> Please be specific (or better, show code):
> 
> * How did you obtain GeometricElementProps - `IModelConnection.elements.getProps()`?
> * What did you try to pass to `IModelJsonReader.parse()` - `GeometricElementProps.geom`?
> * What do you mean by "it failed"? Did it produce exceptions - if so, supply them.
> 
> Did you check if your `GeometricElementProps` object's `geom` member was `undefined`?
> `IModelConnection.elements.getProps()` omits the geometry because (1) it can be an extremely large amount of data that (2) most callers don't care about and (3) only a subset of it can be processed on the front-end.
> 
> What specifically are you trying to accomplish?

Hi, I want to get some geometric attributes of geometric elements in BIM, such as the diameter or height of a column, and then create some decorative effects based on the data.

At present, I have implemented this function with the following main code. Maybe the process is not perfect. My general idea is to query its geometric attributes from the backEnd according to the element ID, and then

This attribute data is used to create a corresponding geometry object in the frontEnd, so we can use the API in the geometry library to achieve more functions.

//User defined RPC interface for obtaining geometric properties 
 const ps = await PropertiesRpcInterface.getClient().getElementProperties(
    imodel.getRpcProps(),
    id,
    true,
    true
  );
  //This part of the code attempts to implement from the interface level, failed, 
  //and then can only rely on the user to determine the type to achieve, the code looks ugly.
    if (ps as GeometricElementProps) {
    const geomElem = ps as GeometricElementProps;
    const geom = geomElem.geom!;
	
    const prop: GeometryStreamEntryProps = geom[1];
    const shapeProp: IModelJson.TorusPipeProps = {
      ...(prop as any).value,
    };
    if (shapeProp) {
	//Successfully created the required geometric objects
	  const torusPipe = IModelJson.Reader.parseTorusPipe(shapeProp);
    }
  }

**Reply by pmconne** [reply]
There is currently no way to obtain an element's geometry in a usable format from the frontend.
[IModelReadRpcInterface.getMassProperties](https://github.com/imodeljs/imodeljs/blob/780dbad115310fcdfbdc921882b0d6e0f29fa903/core/common/src/rpc/IModelReadRpcInterface.ts#L80) can give you various geometric attributes of an element, like [length, volume, perimeter, etc](https://github.com/imodeljs/imodeljs/blob/780dbad115310fcdfbdc921882b0d6e0f29fa903/core/common/src/MassProperties.ts#L35).
If your element is a "column", does it not expose interesting properties you can query directly, like its height and diameter?

**Reply by shenshengyi** [reply]
> There is currently no way to obtain an element's geometry in a usable format from the frontend.
> [IModelReadRpcInterface.getMassProperties](https://github.com/imodeljs/imodeljs/blob/780dbad115310fcdfbdc921882b0d6e0f29fa903/core/common/src/rpc/IModelReadRpcInterface.ts#L80) can give you various geometric attributes of an element, like [length, volume, perimeter, etc](https://github.com/imodeljs/imodeljs/blob/780dbad115310fcdfbdc921882b0d6e0f29fa903/core/common/src/MassProperties.ts#L35).
> If your element is a "column", does it not expose interesting properties you can query directly, like its height and diameter?



OK, thank you very much for your answer. It helped me a lot.
